### PR TITLE
fix: prevent ../ in application path (#9964)

### LINKED ISSUE #9964: Prevent `../`in Application `path` on POST
# Summary

The UI validates the Application's `path` field before creating a new Application. But if someone directly POSTed to the application endpoint, they could include `../` in the `path`.

Actual traversal would still be blocked in `GenerateManifest`. But an extra layer of protection would be good.

# Motivation

If some part of Argo CD code trusted instead of validating the `path` field, someone might be able to pull off a path traversal.

# Proposal

In server/application.go, in the `Create` method, check for traversal before allowing the Application to be created. So the same in the Update method.

Obviously if someone had direct k8s API write access, they could just set the `path` field to whatever they wanted. This check only blocks the invalid path for people who only have Argo CD API access.

### PR DESCRIPTION
**Pull request by notfromstatefarm**
Signed-off-by: notfromstatefarm <86763948+notfromstatefarm@users.noreply.github.com>

Note on DCO:

If the DCO action in the integration test fails, one or more of your commits are not signed off. Please click on the *Details* link next to the DCO action for instructions on how to resolve this.

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/tree/master/community#contributing-to-argo)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)). 

Fixes #9964

### CODE REVIEW

File: util/argo/argo.go [hunk 1]
```diff
@@ -135,6 +136,29 @@ func TestRepoWithKnownType(ctx context.Context, repoClient apiclient.RepoServerS
 	return err
 }
 
+// ValidatePath validates the path specified in application spec. It is checked to ensure that
+// it does not leave the bounds of the repo
+func ValidatePath(path string) error {
+	cleanPath := filepath.Clean(path)
+	if filepath.IsAbs(cleanPath) {
+		return fmt.Errorf("application path is absolute")
```
Comment (crenshaw-dev) [inline util/argo/argo.go:144]: I think we actually have logic elsewhere to interpret absolute paths as being relative to the repo root. I gotta double-check though.

**Review by crenshaw-dev — CHANGES_REQUESTED** [review]
Need to follow up on one thing.  :-)

**Comment by crenshaw-dev** [comment]
Would you be willing to write a unit test from the top level as well, i.e. trying to create/update the app?

**Comment by notfromstatefarm** [comment]
I shall try! Tests are my weak spot. 😄

**Comment by crenshaw-dev** [comment]
Haha mine as well. If I can't copy/paste it, I'm in a tough spot. :-P

**Comment by crenshaw-dev** [comment]
Closing for now since I don't think either Jake or I have the time to tackle it now. Can always reopen.

**Pull request closed without merging**